platform/broadcom/sonic-platform-modules-ufispace/s9301-32db/modules/x86-64-ufispace-s9301-32db-sys-eeprom.c

```diff
@@ -0,0 +1,283 @@
+{
+    struct i2c_client *client = to_i2c_client(container_of(kobj, struct device, kobj));
+    struct eeprom_data *data = i2c_get_clientdata(client);
+    int ret;
+    int i;
+    u8 cmd;
+    u16 value16;
+
+    dev_dbg(&client->dev, "sys_eeprom_write off=%d, count=%d\n", (int)off, (int)count);
+
+    if (off > EEPROM_SIZE) {
+        return 0;
+    }
+    if (off + count > EEPROM_SIZE) {
+        count = EEPROM_SIZE - off;
+    }
+    if (count == 0) {
+        return 0;
+    }
+
+    mutex_lock(&data->update_lock);
+
+    for(i=0; i < count; i++) {
+        /* write command */
+        cmd = (off >> 8) & 0xff;
+        value16 = off & 0xff;
+        value16 |= buf[i] << 8;
+        ret = i2c_smbus_write_word_data(client, cmd, value16);
+
+        if (ret < 0) {
+            dev_err(&client->dev, "write address failed at %d \n", (int)off);
+            goto exit;
+        }
+
+        off++;
+        
+        /* need to wait for write complete */
+        udelay(10000);
+    }
+exit:
+    mutex_unlock(&data->update_lock);
+    /* force to update client when reading */
+    for(i=0; i < SLICE_NUM; i++) {
+        data->last_updated[i] = 0;
+    }
+
+    return count;
+}
+
+static struct bin_attribute sys_eeprom_attr = {
+    .attr = {
+        .name = "eeprom",
+        .mode = S_IRUGO | S_IWUSR,
+    },
+    .size = EEPROM_SIZE,
+    .read = sys_eeprom_read,
+    .write = sys_eeprom_write,
+};
+
+/* Return 0 if detection is successful, -ENODEV otherwise */
+static int sys_eeprom_detect(struct i2c_client *client, struct i2c_board_info *info)
+{
+    struct i2c_adapter *adapter = client->adapter;
+
+    /* EDID EEPROMs are often 24C00 EEPROMs, which answer to all
+       addresses 0x50-0x57, but we only care about 0x51 and 0x55. So decline
+       attaching to addresses >= 0x56 on DDC buses */
+    if (!(adapter->class & I2C_CLASS_SPD) && client->addr >= 0x56) {
+        return -ENODEV;
+    }
+
+    if (!i2c_check_functionality(adapter, I2C_FUNC_SMBUS_READ_BYTE)
+     && !i2c_check_functionality(adapter, I2C_FUNC_SMBUS_WRITE_BYTE_DATA)) {
+        return -ENODEV;
+    }
+
+    strlcpy(info->type, "eeprom", I2C_NAME_SIZE);
+
+    return 0;
+}
+
+static int sys_eeprom_probe(struct i2c_client *client,
+            const struct i2c_device_id *id)
+{
+    struct eeprom_data *data;
+    int err;
+
+    if (!(data = kzalloc(sizeof(struct eeprom_data), GFP_KERNEL))) {
+        err = -ENOMEM;
+        goto exit;
+    }
+
+#ifdef __STDC_LIB_EXT1__
+    memset_s(data->data, EEPROM_SIZE, 0xff, EEPROM_SIZE);
+#else
+    _memset(data->data, 0xff, EEPROM_SIZE);
+#endif
+
+    i2c_set_clientdata(client, data);
+    mutex_init(&data->update_lock);
+
+    /* create the sysfs eeprom file */
+    err = sysfs_create_bin_file(&client->dev.kobj, &sys_eeprom_attr);
+    if (err) {
+        goto exit_kfree;
+    }
+
+    return 0;
+
+exit_kfree:
+    kfree(data);
+exit:
+    return err;
+}
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 1, 0)
```

Comment:
For master branch, kernel modules won't be compiled with anything older than the 6.1 kernel, so I don't think the version check is needed here. Or is this code shared with other things?

Reply:
Yes, we want to share the same driver code across all SONiC branches and UfiSpace's internal projects.